arrow-flight/src/encode.rs

```diff
@@ -194,17 +211,28 @@ impl FlightDataEncoder {
             app_metadata: Some(app_metadata),
             queue: VecDeque::new(),
             done: false,
+            descriptor: None,
         };
 
         // If schema is known up front, enqueue it immediately
         if let Some(schema) = schema {
             encoder.encode_schema(&schema);
         }
+
+        encoder.descriptor = descriptor;
+
         encoder
     }
 
     /// Place the `FlightData` in the queue to send
     fn queue_message(&mut self, data: FlightData) {
+        if let Some(descriptor) = &self.descriptor {
+            let mut data = data;
+            data.flight_descriptor = Some(descriptor.clone());
+            self.descriptor = None;
+            self.queue.push_back(data);
+            return;
+        }
         self.queue.push_back(data);
     }
```

Comment:
FWIW I think you can write this using `take()` a bit more concisely like:

```suggestion
    fn queue_message(&mut self, mut data: FlightData) {
        if let Some(descriptor) = self.descriptor.take() {
            data.flight_descriptor = Some(descriptor.clone());
        }
        self.queue.push_back(data);
    }
```